molviewspec/molviewspec/utils.py

```diff
@@ -62,3 +70,271 @@ def get_major_version_tag() -> str:
+
+    # Convert to Path objects for easier path manipulation
+    input_path = Path(input_mvsj_path)
+    output_path = Path(output_mvsx_path)
+
+    # Read the input MVSJ file
+    try:
+        with open(input_path, 'r', encoding='utf-8') as f:
+            mvsj_data = json.load(f)
+    except Exception as e:
+        logger.error(f"Error reading MVSJ file: {e}")
+        return False
+
+    # Create a temporary directory for downloaded files
+    with tempfile.TemporaryDirectory() as temp_dir:
+        # Find all URI references
+        uri_references = set()
+
+        # Start from root node
+        if "root" in mvsj_data:
+            find_uri_references(mvsj_data["root"], uri_references)
+
+        # If it's a multi-state file, check each snapshot
+        if "snapshots" in mvsj_data and isinstance(mvsj_data["snapshots"], list):
+            for snapshot in mvsj_data["snapshots"]:
+                if "root" in snapshot:
+                    find_uri_references(snapshot["root"], uri_references)
+
+        # Create a mapping from original URIs to archive paths
+        uri_mapping = {}
+
+        # Process each URI reference
+        for uri in uri_references:
+            try:
+                parsed_uri = urlparse(uri)
+
+                # Check if it's an external URI that needs to be downloaded
+                is_external = parsed_uri.scheme in ('http', 'https', 'ftp')
+
+                if is_external and download_external:
+                    # Form the full URL
+                    if base_url and not parsed_uri.netloc:
+                        # Relative URL with base_url provided
+                        full_url = urljoin(base_url, uri)
+                    else:
+                        full_url = uri
+
+                    logger.info(f"Downloading external resource: {full_url}")
+
+                    # Extract filename from URL path
+                    filename = os.path.basename(parsed_uri.path)
+                    if not filename:
+                        # Generate a name if none is available
+                        filename = f"resource_{len(uri_mapping)}"
+
+                    # Ensure we have a unique filename
+                    local_path = os.path.join(temp_dir, filename)
+                    counter = 1
+                    while os.path.exists(local_path):
+                        name, ext = os.path.splitext(filename)
+                        filename = f"{name}_{counter}{ext}"
+                        local_path = os.path.join(temp_dir, filename)
+                        counter += 1
+
+                    # Download the file
+                    try:
+                        urllib.request.urlretrieve(full_url, local_path)
+                        # Add to the mapping
+                        uri_mapping[uri] = filename
+                    except Exception as e:
+                        logger.error(f"Failed to download {full_url}: {e}")
```

Comment:
Would raising the exception be more appropriate here? Now, it seems one can end up with an archive missing files.
Same for the local file path below.